src/backlight/labelizer/ternary/dynamic_neutral.py

```diff
@@ -46,3 +46,33 @@ def neutral_window(self) -> str:
     @property
     def neutral_hard_limit(self) -> str:
         return self._params["neutral_hard_limit"]
+
+
+class MarketCloseAwareDynamicNeutralLabelizer(DynamicNeutralLabelizer):
+    def _calculate_dynamic_neutral_range(self, diff_abs: pd.Series) -> pd.Series:
+
+        df = pd.DataFrame(diff_abs, columns=["res"])
+        df.loc[:, "est"] = df.index.tz_convert("America/New_York")
+        freq = int(
+            pd.Timedelta(self._params["neutral_window"])
+            / pd.Timedelta(diff_abs.index.freq)
+        )
+
+        mask = (
+            ~((df.est.dt.hour <= 17) & (df.est.dt.dayofweek == 6))
+            & ((df.est.dt.hour < 16) | (df.est.dt.hour > 17))
+            & ~((df.est.dt.hour >= 16) & (df.est.dt.dayofweek == 4))
+            & (df.est.dt.dayofweek != 5)
+        )
+
+        dnr = (
+            df.loc[mask, "res"]
+            .rolling(freq)
+            .quantile(self.neutral_ratio)
+            .reindex(diff_abs.index)
+            .fillna(100)
```

Comment:
I couldn't understand this line `fillna(100)`.

Reply:
Thank you for reviewing!

If we remove the `fillna()`, we'll get labels with some `NaN` in the neutral range during NYK close hours as following.

![Screenshot 2021-02-05 at 8 20 17](https://user-images.githubusercontent.com/24562895/106967929-47580d00-678b-11eb-9143-6a311a1ef52d.png)

And that means the labelizer always generates either UP or DOWN for NYK close even if the `label_diff` is small enough. I personally think we should suppress the signals even though the signals for NYK close are currently out of the scope for our current project.

![Screenshot 2021-02-05 at 8 37 45](https://user-images.githubusercontent.com/24562895/106969042-71aaca00-678d-11eb-8173-b16afe1bdc22.png)

We can change from `100` to any numbers if they're big enough.

Reply:
Or shall we forward-fill the `Nan`?
It can possibly make more sense, given that this labelizer aims to capture the market momentum in the recent sessions.

Reply:
Thanks for explanations! I agree with you that ffill may make more sense.